### PR TITLE
Issue 5631. Fix mkdir and is_dir race condition when preparing directory

### LINKED ISSUE #5631: Server 500 error on image resource for Promoted Card in view.
## Description of the bug

Occasionally after installing a new site the front page does not render properly. See:

image

Clearing the cache resolves the problem AND I'm quite sure that it doesn't happen every time. 

NOTE: This occurred when spinning up the site in Lando. Immediately afterwards, I destroyed the site and rebuilt it and it loaded as expected No problem.

## Steps To Reproduce

I can not provide reliable steps to reproduce. 

## Proposed Solutions

Do we need to add another clearing of the cache to the install process. It seems like maybe something is not loading fast enough, under some uncertain conditions.

WILD SPECULATION: 
The fact that the third image is not displaying properly suggests to me that maybe the page is rendering before that image is ready to be rendered.

## Additional Information

I believe that I have only seen this when spinning up a site with Lando. Could that be a factor?

In Zulip @argiepiano mentioned having seen this problem before as well. But, if this is not a regular problem or even a frequent problem, I'm not sure if we should worry about it.

### PR DESCRIPTION
**Pull request by argiepiano**
Fixes https://github.com/backdrop/backdrop-issues/issues/5631

### CODE REVIEW

File: core/includes/file.inc [hunk 1]
```diff
@@ -456,7 +456,7 @@ function file_prepare_directory(&$directory, $options = FILE_MODIFY_PERMISSIONS)
     if (($options & FILE_CREATE_DIRECTORY) && @backdrop_mkdir($directory, NULL, TRUE)) {
       return backdrop_chmod($directory);
     }
-    return FALSE;
+    return is_dir($directory);
```
Comment (indigoxela) [inline core/includes/file.inc]: If I get it right, this might return "true", but doesn't check if it's writable - please correct me if I'm wrong.

Comment (argiepiano) [reply core/includes/file.inc]: Good point. PR updated to check for writable dir

File: core/includes/file.inc [hunk 2]
```diff
@@ -456,7 +456,8 @@ function file_prepare_directory(&$directory, $options = FILE_MODIFY_PERMISSIONS)
     if (($options & FILE_CREATE_DIRECTORY) && @backdrop_mkdir($directory, NULL, TRUE)) {
       return backdrop_chmod($directory);
     }
-    return FALSE;
+    // Check directory existence again @see https://github.com/backdrop/backdrop-issues/issues/5631
```
Comment (quicksketch) [inline core/includes/file.inc]: Let's add a little more explanation and wrap at 80 lines:
```suggestion
    // Race-conditions could lead to the directory having been created by
    // another request. Check if the directory exists again after failing.
    // See https://github.com/backdrop/backdrop-issues/issues/5631
```

File: core/includes/file.inc [hunk 3]
```diff
@@ -456,7 +456,10 @@ function file_prepare_directory(&$directory, $options = FILE_MODIFY_PERMISSIONS)
     if (($options & FILE_CREATE_DIRECTORY) && @backdrop_mkdir($directory, NULL, TRUE)) {
       return backdrop_chmod($directory);
     }
-    return FALSE;
+    // Race-conditions could lead to the directory having been created by
```
Comment (quicksketch) [inline core/includes/file.inc]:
```suggestion
    // Race conditions could lead to the directory having been created by
```

**Comment by quicksketch** [comment]
Merged into 1.x and 1.22.x. Thanks!

**Pull request merged**